cli/src/semgrep/settings.py

```diff
@@ -34,15 +36,41 @@
 
 class SettingsSchema(TypedDict, total=False):
     has_shown_metrics_notification: bool
-    api_token: str
+    api_token: Optional[str]
     anonymous_user_id: str
 
 
 SettingsKeys = Literal[
     "has_shown_metrics_notification", "api_token", "anonymous_user_id"
 ]
 
-DEFAULT_SETTINGS: SettingsSchema = {"anonymous_user_id": str(uuid.uuid4())}
+
+def generate_anonymous_user_id(api_token: Optional[str]) -> str:
+    return (
+        str(uuid.uuid4())
+        if api_token is None
+        else str(
+            uuid.uuid5(
+                uuid.UUID("0" * 32), hashlib.sha256(api_token.encode()).hexdigest()
+            )
+        )
+    )
+
+
+def generate_default_settings(api_token: Optional[str] = None) -> SettingsSchema:
+    anonymous_user_id = generate_anonymous_user_id(api_token)
+    return (
+        {
+            "has_shown_metrics_notification": False,
+            "api_token": api_token,
+            "anonymous_user_id": anonymous_user_id,
+        }
+        if api_token is not None
+        else {
+            "has_shown_metrics_notification": False,
+            "anonymous_user_id": anonymous_user_id,
+        }
+    )
```

Comment:
this might be a little less repetitive if you do something like this

```suggestion
    settings = {
        "has_shown_metrics_notification": False,
        "anonymous_user_id": anonymous_user_id,
    }
    if api_token is not None:
        settings["api_token"] api_token
    return settings
```

Reply:
Yep, I debated whether I wanted to go with the natural, pythonic inclination use a mutation (gasp) to reduce the LOC here – and went with the more "pure" approach.

Reply:
found a better signature – will commit and push!